pkg/workloads/console/runner/runner.go

```diff
@@ -232,30 +232,71 @@ func (c *Runner) Create(ctx context.Context, opts CreateOptions) (*workloadsv1al
 			return csl, err
 		}
 
-		return csl, c.exitWithStatus(csl)
+		return csl, c.waitForSuccess(ctx, csl)
 	}
 
 	return csl, nil
 }
 
-func (c *Runner) exitWithStatus(csl *workloadsv1alpha1.Console) error {
+func (c *Runner) waitForSuccess(ctx context.Context, csl *workloadsv1alpha1.Console) error {
+	isRunning := func(pod *corev1.Pod) bool {
+		return pod != nil && pod.Status.Phase == "Running"
+	}
+
+	succeeded := func(pod *corev1.Pod) bool {
+		return pod != nil && pod.Status.Phase == "Succeeded"
+	}
+
 	pod, _, err := c.GetAttachablePod(csl)
 	if err != nil {
 		return err
 	}
 
-	for pod.Status.Phase == "Running" {
-		pod, _, err = c.GetAttachablePod(csl)
-		if err != nil {
-			return err
-		}
+	listOptions := metav1.SingleObject(pod.ObjectMeta)
+	w, err := c.clientset.CoreV1().Pods(pod.Namespace).Watch(ctx, listOptions)
+	if err != nil {
+		return fmt.Errorf("error watching pod: %w", err)
 	}
 
-	if pod.Status.Phase == "Succeeded" {
+	// We need to fetch the pod again now we have a watcher to avoid a race
+	// where the pod completed before we were listening for watch events
+	pod, _, err = c.GetAttachablePod(csl)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return fmt.Errorf("error retrieving console: %w", err)
+	}
+
+	if succeeded(pod) {
 		return nil
 	}
 
-	return fmt.Errorf("Pod in unsuccessful state %s: %s", pod.Status.Phase, pod.Status.Message)
+	if !isRunning(pod) {
```

Comment:
Since this runs after Create and Attach have completed, this depends on whether a console can be in Running phase while its underlying pod is Pending - I hope not?

Reply:
That's fair - as long as we don't end up calling this function from anywhere else then all is good. Probably not one to worry about for the moment, as any future programming error will manifest itself in an obvious and non-dangerous way